core/sql/optimizer/GroupAttr.cpp

```diff
@@ -580,6 +612,127 @@ void GroupAttributes::addSuitableCompRefOptConstraints(
+    tryToEliminateOrderColumnBasedOnEqualsPred(col, NULL);
+}
+
+// This and the previous method are used to match required sort orders
+// or arrangements to an actual key ordering of a table, in cases
+// where some columns are equated to a constant, like this, with the
+// clustering key being (a,b,c)
+//
+// select ...
+// from t
+// where a = 5
+// order by b
+//
+// Note that for VEGs, this is handled differently (see
+// ValueIdList::satisfiesReqdOrder), this code is only for non-VEG
+// cases (usually computed columns, also varchar).
+//
+// If we eliminate a column based on a supplied predicate, then
+// this predicate is added as an Optimizer check constraint to the
+// group attributes, so that future calls will continue to accept
+// the simplified sort order.
+NABoolean GroupAttributes::tryToEliminateOrderColumnBasedOnEqualsPred(
+     ValueId col,
+     const ValueIdSet *preds)
+{
+  NABoolean result = FALSE;
+
+  if (preds || hasConstraintOfType(ITM_CHECK_OPT_CONSTRAINT))
+    {
+      // Comparison failed. If the caller provided predicates,
+      // then we can try something similar to what we did with
+      // group attributes above. If the predicate equate the
+      // column with a constant, then we can eliminate it as
+      // well. Note that we don't expect the requirements to
+      // omit such columns, since the parent will usually not
+      // know about such predicates, so we check this condition
+      // only after trying the regular method.
+      //
+      // This situation typically happens with computed columns
+      // where predicates are added after VEGs are formed
+
+      ValueIdSet checkConstraints;
+      ValueIdSet checkPreds;
+
+      // look for predicates we remembered from earlier calls
+      getConstraintsOfType(ITM_CHECK_OPT_CONSTRAINT,
+                           checkConstraints);
+
+      for (ValueId c = checkConstraints.init();
+           checkConstraints.next(c);
+           checkConstraints.advance(c))
+        checkPreds += static_cast<CheckOptConstraint *>(
+             c.getItemExpr())->getCheckPreds();
+
+      // also use newly provided predicates
+      if (preds)
+        checkPreds += *preds;
+
+      // if the column is descending, then get rid of the Inverse
+      // operator for the next check
+      if (col.getItemExpr()->getOperatorType() == ITM_INVERSE)
+        col = col.getItemExpr()->child(0).getValueId();
+
+      // convert col from a VEGRef to a base column, if needed,
+      // the ScanKey method below wants a real column as input
+      if (col.getItemExpr()->getOperatorType() == ITM_VEG_REFERENCE)
+        {
+          const ValueIdSet &vegMembers =
+            static_cast<VEGReference *>(col.getItemExpr())->
+            getVEG()->getAllValues();
+          for (ValueId b=vegMembers.init();
+               vegMembers.next(b);
+               vegMembers.advance(b))
+            if (b.getItemExpr()->getOperatorType() == ITM_BASECOLUMN)
+              col = b;
```

Comment:
In the meantime I looked more at the code and tried a few options. Removing this code makes the test case fail. In the commit for the rework, I'll replace the current check (VEG member is a base column) with a check whether the VEG member is referenced by the check constraints. That should ensure that we use the correct value id (I found that the value selected is actually an index column).

About the issue of applying transitive closure incorrectly, when we have a predicate T1.a=const and a VEG(T1.a, T2.b): I'll add a comment to the code explaining why I think this won't happen.